volatility3/framework/layers/qemu.py

```diff
@@ -69,6 +73,117 @@ def _read_configuration(self, base_layer: interfaces.layers.DataLayerInterface,
                     return dict()
         raise exceptions.LayerException(name, "Invalid JSON configuration at the end of the file")
 
+    @staticmethod
+    def _get_pci_hole(architecture: Optional[Tuple[str, Optional[str]]], ram_size: int) -> \
+            Optional[Tuple[int, int]]:
+        """Returns the location and size of the PCI-hole for the given architecture and RAM size.
+
+        Args:
+            architecture: (Optionally) A tuple of strings containing the name (in lowercase) and (optionally) the
+                version of the used QEMU target architecture
+            ram_size: Size of the RAM in bytes
+
+        Returns:
+            (Optionally) A tuple containing the start and size of the PCI hole
+        """
+
+        # See https://qemu.readthedocs.io/en/latest/devel/memory.html for more info
+        #
+        # At least the following values could occur for devices using > 3-4 GB RAM:
+        # +--------------------------------+--------------------------------+------------+-------------+
+        # | Architecture                   | Reference Code                 | Hole Start | Hole End    |
+        # +--------------------------------+--------------------------------+------------+-------------+
+        # | PC i440FX + PIIX "New Default" | qemu/hw/i386/pc_piix.c:98      | 0xc0000000 | 0x100000000 |
+        # | PC i440FX + PIIX "Old Default" | qemu/hw/i386/pc_piix.c:98      | 0xe0000000 | 0x100000000 |
+        # | PC Q35 + ICH9                  | qemu/hw/i386/pc_q35.c:141      | 0x80000000 | 0x100000000 |
+        # | MicroVM                        | qemu/hw/i386/microvm.c:291     | 0xc0000000 | 0x100000000 |
+        # | Xen                            | qemu/hw/i386/xen/xen-hvm.c:248 | 0xf0000000 | 0x100000000 |
+        # +--------------------------------+--------------------------------+------------+-------------+
+        #
+        # MicroVM and Xen haven't been implemented yet
+        #
+        # For now, we assume that the parameter max-ram-below-4g is not set, since this parameter influences the size
+        # and location of the memory gap. Deviating hole sizes could eventually be detected for Linux by e.g. scanning
+        # for dmesg entries with a regex like rb'\[mem (0x[0-9a-f]{4,10})-0x[0-9a-f]{4,10}\] available for PCI devices'
+
+        end = 0x100000000
+
+        if architecture:
```

Comment:
This feels like it's data, and there's not a great reason to have this stuff calculated, rather than in a big look-up table.  I think it would be easier to see what's going on just to have a `Dict[Tuple[...], Tuple[int, int]]`